### ISSUE TITLE
Issues parsing tables with colspans/rowspans

### ISSUE THREAD

**Issue by nschonni** [validator/validator#1069]
@sideshowbarker I've mostly got the `mdn/content` repo green (besides some filtering issues I'll spin up separately)

There appears to be a few issues in the table parsing/calculation when rowspan/colspan are used, and the validator thinks that elements are spilling outside the table.

```
"file:/mnt/z/mdn-content/files/en-us/web/api/filesystemflags/index.html":44.7-45.59: error: Table column 2 established by element “th” has no cells beginning in it.
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":220.10-221.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":722.10-723.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":1113.10-1114.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":1275.10-1276.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":1542.10-1543.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":1870.10-1871.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/api/keyboardevent/keycode/index.html":2141.10-2142.8: error: Element “tbody” not allowed as child of element “table” in this context. (Suppressing further errors from this subtree.)
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/matrix3d()/index.html":49.7-50.31: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/matrix3d()/index.html":51.16-52.19: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/matrix3d()/index.html":54.33-55.19: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/perspective()/index.html":52.9-53.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/perspective()/index.html":55.12-56.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/perspective()/index.html":57.50-58.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/rotatez()/index.html":57.9-58.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/rotatez()/index.html":59.12-60.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/rotatez()/index.html":111.21-112.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scale3d()/index.html":58.9-59.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scale3d()/index.html":60.12-61.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scale3d()/index.html":99.21-100.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scalez()/index.html":58.9-59.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scalez()/index.html":60.12-61.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/scalez()/index.html":99.21-100.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translate3d()/index.html":52.9-53.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translate3d()/index.html":55.12-56.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translate3d()/index.html":57.42-58.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translatez()/index.html":61.9-62.34: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translatez()/index.html":63.12-64.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
"file:/mnt/z/mdn-content/files/en-us/web/css/transform-function/translatez()/index.html":65.42-66.22: error: Table cell spans past the end of its row group established by a “tbody” element; clipped to the end of the row group.
```

**Comment by sideshowbarker** [comment]
> There appears to be a few issues in the table parsing/calculation when rowspan/colspan are used, and the validator thinks that elements are spilling outside the table.

I believe that’s because there’s an error in the table markup, but browsers are correcting for it.

I will make time to look into this, but in the meantime it could help to have, recorded here, a minimal test case that can be used to reproduce the error.

**Comment by nschonni** [comment]
Here is the first table pulled out
```html
<table class="standard-table">
 <thead>
  <tr>
   <th colspan="3" scope="col" style="text-align: center;">Option values</th>
   <th rowspan="2" scope="col">File/directory condition</th>
   <th rowspan="2" scope="col">Result</th>
  </tr>
  <tr>
   <th colspan="2" scope="col">{{domxref("FileSystemFlags.create", "create")}}</th>
   <th scope="col">{{domxref("FileSystemFlags.exclusive", "exclusive")}}</th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td colspan="2"><code>false</code></td>
   <td>n/a<sup>[1]</sup></td>
   <td>Path exists and matches the desired type (depending on whether the function called is <code>getFile()</code> or <code>getDirectory()</code></td>
   <td>The <code>successCallback</code> is called with a {{domxref("FileSystemFileEntry")}} if <code>getFile()</code> was called or a {{domxref("FileSystemDirectoryEntry")}} if <code>getDirectory()</code> was called.</td>
  </tr>
  <tr>
   <td colspan="2"><code>false</code></td>
   <td>n/a<sup>[1]</sup></td>
   <td>Path exists but doesn't match the desired type</td>
   <td>The <code>errorCallback</code> is called with an appropriate error code (if the callback was provided).</td>
  </tr>
  <tr>
   <td colspan="2"><code>true</code></td>
   <td><code>false</code></td>
   <td>Path exists</td>
   <td>The existing file or directory is removed and replaced with a new one, then the <code>successCallback</code> is called with a {{domxref("FileSystemFileEntry")}} or a {{domxref("FileSystemDirectoryEntry")}}, as appropriate.</td>
  </tr>
  <tr>
   <td colspan="2"><code>true</code></td>
   <td><code>false</code></td>
   <td>Path doesn't exist</td>
   <td>The file or directory is created, then a {{domxref("FileSystemFileEntry")}} or a {{domxref("FileSystemDirectoryEntry")}} is passed to the <code>successCallback</code>, as appropriate.</td>
  </tr>
  <tr>
   <td colspan="2"><code>true</code></td>
   <td><code>true</code></td>
   <td>Path exists</td>
   <td>The <code>errorCallback</code> is called with an appropriate error, such as <code>FileError.PATH_EXISTS_ERR</code>.</td>
  </tr>
  <tr>
   <td colspan="2"><code>true</code></td>
   <td><code>true</code></td>
   <td>Path doesn't exist</td>
   <td>The file or directory is created, then a {{domxref("FileSystemFileEntry")}} or a {{domxref("FileSystemDirectoryEntry")}} is passed to the <code>successCallback</code>, as appropriate.</td>
  </tr>
 </tbody>
</table>
```

**Comment by sideshowbarker** [comment]
OK, the rendered table looks like this:

![image](https://user-images.githubusercontent.com/194984/103845321-2f835f80-50df-11eb-93b5-5dd9e852989d.png)

So looking at that just visually and comparing it to the markup, I see:

* the **Option values** cell spans 2 columns, but the markup for it has `colspan="3"`
* the `create` cell spans only 1 column, but its markup has `colspan="2"`
* the `false` cell and all cells in that column span only 1 column, but their markup has `colspan="2"`

So the table markup appears to be incorrect, and the checker is catching the errors and reporting them.

Here’s a diff that shows the corrections needed to the table markup to fix the errors:

```html
$ diff -u0 X.html Y.html
--- X.html	2021-01-07 11:40:50.643237787 +0900
+++ Y.html	2021-01-07 11:41:05.042807862 +0900
@@ -4 +4 @@
-   <th colspan="3" scope="col" style="text-align: center;">Option values</th>
+   <th colspan="2" scope="col" style="text-align: center;">Option values</th>
@@ -9 +9 @@
-   <th colspan="2" scope="col">{{domxref("FileSystemFlags.create", "create")}}</th>
+   <th scope="col">{{domxref("FileSystemFlags.create", "create")}}</th>
@@ -15 +15 @@
-   <td colspan="2"><code>false</code></td>
+   <td><code>false</code></td>
@@ -21 +21 @@
-   <td colspan="2"><code>false</code></td>
+   <td><code>false</code></td>
@@ -27 +27 @@
-   <td colspan="2"><code>true</code></td>
+   <td><code>true</code></td>
@@ -33 +33 @@
-   <td colspan="2"><code>true</code></td>
+   <td><code>true</code></td>
@@ -39 +39 @@
-   <td colspan="2"><code>true</code></td>
+   <td><code>true</code></td>
@@ -45 +45 @@
-   <td colspan="2"><code>true</code></td>
+   <td><code>true</code></td>
```

**Comment by nschonni** [comment]
Figured out the transform-function ones were also correct with bad rowspans and submitted https://github.com/mdn/content/pull/1067
Finally figured out the keyboard one with https://github.com/mdn/content/pull/1068. I still had it in my head that tfoot was supposed to go before body and the message was bad
